src/reflect/scala/reflect/runtime/TwoWayCaches.scala

```diff
@@ -52,7 +54,7 @@ private[runtime] trait TwoWayCaches { self: SymbolTable =>
       }
     }
 
-    def toJavaOption(key: S)(body: => Option[J]): Option[J] = gilSynchronized {
```

Comment:
We need to be sure that calling code was not relying on the coarse grained lock here to safely execute the `body` thunk. We do have a test of runtime reflection thread safety in test/files/run/reflection-sync-potpourri.scala and test/files/run/reflection-sync-subtypes.scala, so it is encouraging to see that they still pass after your patch. But there could be race conditions that aren't exercised by that test.